fireplace/dsl/random_picker.py

```diff
@@ -1,79 +1,96 @@
 import random
 from hearthstone.enums import CardType, Race, Rarity
 from .lazynum import LazyValue
+from copy import deepcopy
+from .. import cards
 
 
 class RandomCardPicker(LazyValue):
 	"""
 	Store filters and generate a random card matching the filters on pick()
+	Constructor takes a single global set of filters and uses a default weighting of 1
+	Additional weighted filter sets can be added with add(), these will be merged with the global filters
 	"""
-	def __init__(self, *args, **filters):
-		self.args = args
-		self.filters = filters
+	def __init__(self, **filters):
+		self.weights = []
+		self.weightedfilters = []
+		self.globalfilter = filters
 		self.count = 1
-		self._cards = None
-		self.lazy_filters = False
-		for v in filters.values():
-			if isinstance(v, LazyValue):
-				self.lazy_filters = True
-				break
 
 	def __repr__(self):
-		return "%s(%r)" % (self.__class__.__name__, self.filters)
+		return "%s(%r)" % (self.__class__.__name__, self.globalfilter)
 
+	# select number of cards to fetch
 	def __mul__(self, other):
-		ret = self.__class__(*self.args, **self.filters)
+		ret = deepcopy(self)
```

Comment:
Yes, I think we should avoid mutating the lists within the body of the picker and use `copy` instead. Make copies of the lists within the body of `evaluate` instead.

For example, right now there's no way to re-use the weighted picker because we never reset the weights/lists after a single call to `evaluate`. That's probably not a serious issue, because we don't attempt to evaluate pickers more than once AFAICT, but then you'd need to document properly that evaluate can be called at most once on pickers, and it pushes the onus of copying the picker onto the user, which I'm not a huge fan of.